### PR TITLE
Refactor measurement visitors

### PR DESCRIPTION
**Pull request by makr11st**
## Proposed changes

* Remove FlatMeasurementVisitor in favour of GroupedMeasurementVisitor.
  Either one can be implemented by the other. Rename to just MeasurementVisitor.

  Grouping semantics (whether or not a group can be reopened for instance) depends on the concrete implementation not on the Visitor trait.

* Split MeasurementGrouper according to separation of concerns principle.

* Prefix visitor methods with "visit_". `serde` visitors use that pattern and generally, methods should name an "action".

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactorings that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: mapper/tedge_mapper/src/collectd_mapper/batcher.rs [hunk 1]
```diff
@@ -39,22 +40,22 @@ impl MessageBatch {
...
-        self.message_grouper.measurement(
-            Some(collectd_message.metric_group_key),
-            collectd_message.metric_key,
-            collectd_message.metric_value,
-        )?;
+        self.message_grouper
+            .visit_start_group(&collectd_message.metric_group_key)?;
+        self.message_grouper
+            .visit_measurement(collectd_message.metric_key, collectd_message.metric_value)?;
+        self.message_grouper.visit_end_group()?;
```
Comment (didier-wenzek) [inline mapper/tedge_mapper/src/collectd_mapper/batcher.rs:47]: This might be simplified by adding an helper method on the `MeasurementVisitor`.

This help method would have the same signature of the former `FlatMeasurementVisitor.measurement()` and performing this `star_group; visit_measurement; end_group` sequence.

File: mapper/thin_edge_json/src/json.rs [hunk 2]
```diff
@@ -141,20 +141,14 @@ pub fn parse_str<T: GroupedMeasurementVisitor>(
     match &thin_edge_obj {
         JsonValue::Object(thin_edge_obj) => {
             for (key, value) in thin_edge_obj.iter() {
-                if key.contains('\\') {
-                    return Err(ThinEdgeJsonError::InvalidThinEdgeJsonKey {
-                        key: String::from(key),
-                    }
-                    .into());
-                }
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/json.rs:149]: Why this has been removed?

To be honest, I have no clue on the purpose of this very specific check, but removing it seems unrelated to that PR.

File: mapper/thin_edge_json/src/measurement.rs [hunk 3]
```diff
@@ -164,19 +62,19 @@ pub trait FlatMeasurementVisitor {
...
-pub trait GroupedMeasurementVisitor {
-    /// Error type specific to this way of collecting measurements
+pub trait MeasurementVisitor {
+    /// Error type specific to this visitor.
     type Error: std::error::Error + std::fmt::Debug;
 
-    /// Set the timestamp shared by all the measurements of this serie
-    fn timestamp(&mut self, value: DateTime<FixedOffset>) -> Result<(), Self::Error>;
+    /// Set the timestamp shared by all the measurements of this series.
+    fn visit_timestamp(&mut self, value: DateTime<FixedOffset>) -> Result<(), Self::Error>;
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/measurement.rs:70]: This is a matter of taste, but I prefer the name `timestamp` without the `visit` prefix!

What matters is that the team is comfortable with the naming,

File: mapper/thin_edge_json/src/measurement.rs [hunk 4]
```diff
@@ -164,19 +62,19 @@ pub trait FlatMeasurementVisitor {
 ///     }
 /// }
 /// ```
-pub trait GroupedMeasurementVisitor {
-    /// Error type specific to this way of collecting measurements
+pub trait MeasurementVisitor {
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/measurement.rs:65]: Why not an helper method that combines the sequence start / measurement / end.

**Review by didier-wenzek — APPROVED** [review]
Approved. I just have some minor comments.

File: mapper/tedge_mapper/src/collectd_mapper/batcher.rs (hunk 1 above)
Comment (mneumann) [reply mapper/tedge_mapper/src/collectd_mapper/batcher.rs:47]: Yes, an extension trait!

File: mapper/thin_edge_json/src/measurement.rs (hunk 4 above)
Comment (mneumann) [reply mapper/thin_edge_json/src/measurement.rs:65]: We can add that to `MeasurementVisitorExt`.

File: mapper/thin_edge_json/src/measurement.rs (hunk 3 above)
Comment (mneumann) [reply mapper/thin_edge_json/src/measurement.rs:70]: IMHO, it makes it clearer what are the callback functions and what are functions that do not belong to the Visitor trait. For instance, the `done` or `finish` methods are not defined in the trait, and these are the methods that should be called directly by the consumer, whereas the callback functions should not be called directly.

Comment (mneumann) [reply mapper/thin_edge_json/src/measurement.rs:70]: NOTE that `serde` also prefixes all visitor methods with `visit_` and I can point you at various other implementations that do exactly the same. I understand it's mostly taste, but there is some further reasoning behind these changes, mainly that methods should name actions.

**Review by mneumann — APPROVED** [review]
I am biased about these changes. Please merge so we can move forward. 

I'll consider adding an `MeasurementVisitorExt` trait that contains the helper method Didier suggested.
